### PR TITLE
rosconsole format for rospy

### PR DESCRIPTION
**Pull request by mjyc**
quick & dirty solution for issue:
https://github.com/ros/ros_comm/issues/517

Since in python it uses [logging module](https://docs.python.org/2/library/logging.html) and in C++ it uses [log4cxx](http://logging.apache.org/log4cxx/index.html), I could not find 1 to 1 translation for some outputs. For example, `${thread}` and `${logger}` does not match well with that of C++.

Also I was not sure if this file was a good place to write this functionality. @dirk-thomas recommended to check out [this file](https://github.com/ros/ros_comm/blob/fab4840ef79f5334bcd043e057a1909183842159/clients/rospy/src/rospy/core.py#L119) but this file was easier to make change. Let me know if you have suggestions,

### CODE REVIEW

File: tools/rosgraph/src/rosgraph/roslogging.py [hunk 1]
```diff
@@ -152,11 +152,26 @@ def __init__(self, colorize=True):
...
+            msg = os.environ['ROSCONSOLE_FORMAT']
+            msg = msg.replace('${severity}', level)
+            msg = msg.replace('${message}', str(record.getMessage()))
+            msg = msg.replace('${time}', str(time.time()))
+            msg = msg.replace('${thread}', str(record.thread))
+            msg = msg.replace('${logger}', str(record.name))
+            msg = msg.replace('${file}', str(record.pathname))
+            msg = msg.replace('${line}', str(record.lineno))
+            msg = msg.replace('${function}', str(record.funcName))
+            from rospy import get_name
```
Comment (dirk-thomas) [inline tools/rosgraph/src/rosgraph/roslogging.py]: This is not possible. You can't import `rospy` here - `ros_graph` does _not_ depend on `rospy`. It can also not add that dependency since `rospy` actually depends on `rosgraph`.

Comment (mjyc) [inline tools/rosgraph/src/rosgraph/roslogging.py]: What if I wrap the code using rospy with try & catch block? Similar to the [line 145-151 of this file](https://github.com/ros/ros_comm/blob/indigo-devel/tools/rosgraph/src/rosgraph/roslogging.py#L145-L151)? Is there some special problem with using `get_name` function of rospy? I tried to modify [core.py](https://github.com/ros/ros_comm/blob/indigo-devel/clients/rospy/src/rospy/core.py) file, for example somehow try to overwrite this `emit` function from `core.py`.. would that be right direction? any other suggestions?

File: tools/rosgraph/src/rosgraph/roslogging.py [hunk 2]
```diff
@@ -152,11 +152,29 @@ def __init__(self, colorize=True):
...
+            msg = msg.replace('${file}', str(record.pathname))
+            msg = msg.replace('${line}', str(record.lineno))
+            msg = msg.replace('${function}', str(record.funcName))
+            try:
+                from rospy import get_name
+                msg = msg.replace('${node}', get_name())
+            except ImportError:
+                msg = msg.replace('${node}', 'unavailable')
+            if self._get_time is not None and not self._is_wallclock():
+                msg += ' [%f]' % self._get_time()
```
Comment (dirk-thomas) [inline tools/rosgraph/src/rosgraph/roslogging.py]: I find it rather unintuitive to append the time at the end of the string. Would an additional replacement of something like `walltime` make more sense?

File: tools/rosgraph/src/rosgraph/roslogging.py [hunk 3]
```diff
@@ -152,11 +152,29 @@ def __init__(self, colorize=True):
...
+            msg = msg.replace('${thread}', str(record.thread))
+            msg = msg.replace('${logger}', str(record.name))
+            msg = msg.replace('${file}', str(record.pathname))
+            msg = msg.replace('${line}', str(record.lineno))
+            msg = msg.replace('${function}', str(record.funcName))
+            try:
+                from rospy import get_name
+                msg = msg.replace('${node}', get_name())
+            except ImportError:
+                msg = msg.replace('${node}', 'unavailable')
```
Comment (dirk-thomas) [inline tools/rosgraph/src/rosgraph/roslogging.py]: Please reduce the redundant code part and only do the minimal stuff inside the try block:

```
try:
    from rospy import get_name
    node_name = get_name()
except ImportError:
    node_name = '<unknown_node_name>'
msg = msg.replace('${node}', node_name)
```

File: tools/rosgraph/src/rosgraph/roslogging.py (hunk 2 above)
Comment (mjyc) [inline tools/rosgraph/src/rosgraph/roslogging.py]: @dirk-thomas  I agree with you on the time at the end of the string. I kept that to be consistent with [the original `msg` string generated by `emit` function](https://github.com/ros/ros_comm/blob/indigo-devel/tools/rosgraph/src/rosgraph/roslogging.py#L153-L158).

Do we need to see both [time.time()](https://github.com/ros/ros_comm/blob/indigo-devel/tools/rosgraph/src/rosgraph/roslogging.py#L153-L158) and [self._get_time()](https://github.com/ros/ros_comm/blob/indigo-devel/tools/rosgraph/src/rosgraph/roslogging.py#L157) If not, how about this emit function?

```
    def emit(self, record):
        level, color = _logging_to_rospy_names[record.levelname]
        if self._get_time is not None and not self._is_wallclock():
            walltime = '[%f]' % self._get_time()
        else:
            walltime = '[%f]' % time.time()
        if 'ROSCONSOLE_FORMAT' in os.environ.keys():
            msg = os.environ['ROSCONSOLE_FORMAT']
            msg = msg.replace('${severity}', level)
            msg = msg.replace('${message}', str(record.getMessage()))
            msg = msg.replace('${time}', walltime)
            msg = msg.replace('${thread}', str(record.thread))
            msg = msg.replace('${logger}', str(record.name))
            msg = msg.replace('${file}', str(record.pathname))
            msg = msg.replace('${line}', str(record.lineno))
            msg = msg.replace('${function}', str(record.funcName))
            try:
                from rospy import get_name
                node_name = get_name()
            except ImportError:
                node_name = '<unknown_node_name>'
            msg = msg.replace('${node}', node_name)
            msg += '\n'
        else:
            msg = '[%s] [WallTime: %f]' % (level, walltime)
            msg += ' %s\n' % record.getMessage()
        if record.levelno < logging.WARNING:
            self._write(sys.stdout, msg, color)
        else:
            self._write(sys.stderr, msg, color)
```

Basically, it tries to use the timestamp from `self._get_time()` first, and if it is not available it resorts to `time.time()`.

Comment (dirk-thomas) [inline tools/rosgraph/src/rosgraph/roslogging.py]: I think it would be more flexible to expose both times through different variables. That way the user can choose what he wants to see (e.g. even both).

**Comment by mjyc** [comment]
@dirk-thomas introduced `${walltime}` variable in addition to the exisiting `${time}` variable as you recommended. FYI, `${walltime}` is `time.time()` to be consistent with [the original code](https://github.com/ros/ros_comm/blob/indigo-devel/tools/rosgraph/src/rosgraph/roslogging.py#L155).

**Comment by dirk-thomas** [comment]
@ros-pull-request-builder

**Comment by dirk-thomas** [comment]
The one failing test is due to the dependency `genpy` not yet being rereleased.
